lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -2744,6 +2746,200 @@ def calc_arrow(uvw, angle=15):
 
     quiver3D = quiver
 
+    def voxels(self, *args, **kwargs):
+        """
+        ax.voxels([x, y, z,] /, filled, **kwargs)
+
+        Plot a set of filled voxels
+
+        All voxels are plotted as 1x1x1 cubes on the axis, with filled[0,0,0]
+        placed with its lower corner at the origin. Occluded faces are not
+        plotted.
+
+        .. versionadded:: 2.1
+
+        Parameters
+        ----------
+        filled : 3D np.array of bool
+            A 3d array of values, with truthy values indicating which voxels
+            to fill
+
+        x, y, z : 3D np.array, optional
+            The coordinates of the corners of the voxels. This should broadcast
+            to a shape one larger in every dimension than the shape of `filled`.
+            These arguments can be used to plot non-cubic voxels.
+            If not specified, defaults to increasing integers along each axis.
+
+        facecolors, edgecolors : array_like
+            The color to draw the faces and edges of the voxels. This parameter
+            can be:
+
+              - A single color value, to color all voxels the same color. This
+                can be either a string, or a 1D rgb/rgba array
+              - ``None``, indicating to use the default
+              - A 3D ndarray of color names, with each item the color for the
+                corresponding voxel. The size must match the voxels.
+              - A 4D ndarray of rgb/rgba data, with the components along the
+                last axis.
+
+        Any additional keyword arguments are passed onto
+        :func:`~mpl_toolkits.mplot3d.art3d.Poly3DCollection`
+
+        Returns
+        -------
+        faces : dict
+            A dictionary indexed by coordinate, where ``faces[i,j,k]`` is a
+            `Poly3DCollection` of the faces drawn for the voxel
+            ``filled[i,j,k]``. If no faces were drawn for a given voxel, either
+            because it was not asked to be drawn, or it is fully occluded, then
+            ``(i,j,k) not in faces``.
+
+        Examples
+        --------
+
+        .. plot:: gallery/mplot3d/voxels.py
+        .. plot:: gallery/mplot3d/voxels_rgb.py
+        .. plot:: gallery/mplot3d/voxels_torus.py
+        """
+
+        # parse optional x y z arguments
+        if len(args) >= 3:
+            xyz = args[:3]
+            args = args[3:]
+        else:
+            xyz = None
+
+        # clever trick to produce the right error message
+        _get_args = lambda filled, **kwargs: (filled, kwargs)
+        _get_args.__name__ = str('voxels')
+        filled, kwargs = _get_args(*args, **kwargs)
```

Comment:
This trick makes it possible to pass filled by kwarg or positional arg, as well as producing sane error messages for missing arguments. The str is needed because __name__ has to match the python "string-literal" type, but this file uses unicode_literals.